File: client/src/main/scala/stasis/client/api/clients/DefaultServerApiEndpointClient.scala
package stasis.client.api.clients

import java.time.Instant

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.control.NonFatal

import org.apache.pekko.Done
import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.http.scaladsl.marshalling.Marshal
import org.apache.pekko.http.scaladsl.model._
import org.apache.pekko.http.scaladsl.model.headers.HttpCredentials
import org.apache.pekko.http.scaladsl.unmarshalling.Unmarshal
import org.apache.pekko.util.ByteString
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import stasis.client.api.clients.exceptions.ServerApiFailure
import stasis.client.encryption.Decoder
import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.model.DatasetMetadata
import stasis.core.api.PoolClient
import stasis.core.networking.exceptions.ClientFailure
import stasis.layers.security.tls.EndpointContext
import stasis.layers.streaming.Operators.ExtendedSource
import stasis.shared.api.requests.CreateDatasetDefinition
import stasis.shared.api.requests.CreateDatasetEntry
import stasis.shared.api.requests.ResetUserPassword
import stasis.shared.api.responses.CreatedDatasetDefinition
import stasis.shared.api.responses.CreatedDatasetEntry
import stasis.shared.api.responses.Ping
import stasis.shared.api.responses.UpdatedUserSalt
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.model.devices.Device
import stasis.shared.model.schedules.Schedule
import stasis.shared.model.users.User

class DefaultServerApiEndpointClient(
  apiUrl: String,
  credentials: => Future[HttpCredentials],
  decryption: DefaultServerApiEndpointClient.DecryptionContext,
  override val self: Device.Id,
  override protected val context: Option[EndpointContext],
  override protected val config: PoolClient.Config
)(implicit override protected val system: ActorSystem[Nothing])
    extends ServerApiEndpointClient
    with PoolClient {
  import DefaultServerApiEndpointClient._
  import stasis.shared.api.Formats._

  override protected val log: Logger = LoggerFactory.getLogger(this.getClass.getName)

  private implicit val ec: ExecutionContext = system.executionContext

  override val server: String = apiUrl

  override def createDatasetDefinition(request: CreateDatasetDefinition): Future[CreatedDatasetDefinition] =
    if (request.device == self) {
      for {
        entity <- request.toEntity
        credentials <- credentials
        response <- offer(
          request = HttpRequest(
            method = HttpMethods.POST,
            uri = s"$apiUrl/v1/datasets/definitions/own",
            entity = entity
          ).addCredentials(credentials = credentials)
        ).transformClientFailures()
        created <- response.to[CreatedDatasetDefinition]
      } yield {
        created
      }
    } else {
      Future.failed(
        new IllegalArgumentException(
          s"Cannot create dataset definition for a different device: [${request.device.toString}]"
        )
      )
    }

  override def createDatasetEntry(request: CreateDatasetEntry): Future[CreatedDatasetEntry] =
    for {
      entity <- request.toEntity
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.POST,
          uri = s"$apiUrl/v1/datasets/entries/own/for-definition/${request.definition.toString}",
          entity = entity
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      created <- response.to[CreatedDatasetEntry]
    } yield {
      created
    }

  override def datasetDefinitions(): Future[Seq[DatasetDefinition]] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/datasets/definitions/own"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      definitions <- response.to[Seq[DatasetDefinition]]
    } yield {
      definitions.filter(_.device == self)
    }

  override def datasetEntries(definition: DatasetDefinition.Id): Future[Seq[DatasetEntry]] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/datasets/entries/own/for-definition/${definition.toString}"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      entries <- response.to[Seq[DatasetEntry]]
    } yield {
      entries
    }

  override def datasetDefinition(definition: DatasetDefinition.Id): Future[DatasetDefinition] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/datasets/definitions/own/${definition.toString}"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      definition <- response.to[DatasetDefinition]
      _ <-
        if (definition.device == self) {
          Future.successful(())
        } else {
          Future.failed(
            new IllegalArgumentException(
              "Cannot retrieve dataset definition for a different device"
            )
          )
        }
    } yield {
      definition
    }

  override def datasetEntry(entry: DatasetEntry.Id): Future[DatasetEntry] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/datasets/entries/own/${entry.toString}"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      entry <- response.to[DatasetEntry]
    } yield {
      entry
    }

  override def latestEntry(definition: DatasetDefinition.Id, until: Option[Instant]): Future[Option[DatasetEntry]] = {
    val baseUrl = s"$apiUrl/v1/datasets/entries/own/for-definition/${definition.toString}/latest"
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = until match {
            case Some(until) => s"$baseUrl?until=${until.toString}"
            case None        => baseUrl
          }
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      entry <- response match {
        case HttpResponse(StatusCodes.NotFound, _, _, _) => Future.successful(None)
        case _                                           => response.to[DatasetEntry].map(Some.apply)
      }
    } yield {
      entry
    }
  }

  override def publicSchedules(): Future[Seq[Schedule]] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/schedules/public"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      schedules <- response.to[Seq[Schedule]]
    } yield {
      schedules
    }

  override def publicSchedule(schedule: Schedule.Id): Future[Schedule] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/schedules/public/${schedule.toString}"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      schedule <- response.to[Schedule]
    } yield {
      schedule
    }

  override def datasetMetadata(entry: DatasetEntry.Id): Future[DatasetMetadata] =
    for {
      entry <- datasetEntry(entry)
      metadata <- datasetMetadata(entry)
    } yield {
      metadata
    }

  override def datasetMetadata(entry: DatasetEntry): Future[DatasetMetadata] =
    decryption match {
      case DecryptionContext.Default(core, deviceSecret, decoder) =>
        for {
          entryMetadata <- core.pull(crate = entry.metadata)
          entryMetadata <- DatasetMetadata.decrypt(
            metadataCrate = entry.metadata,
            metadataSecret = deviceSecret.toMetadataSecret(metadataCrate = entry.metadata),
            metadata = entryMetadata,
            decoder = decoder
          )
        } yield {
          entryMetadata
        }

      case DecryptionContext.Disabled =>
        Future.failed(
          new IllegalStateException("Cannot retrieve dataset metadata; decryption context is disabled")
        )
    }

  override def user(): Future[User] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/users/self"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      user <- response.to[User]
    } yield {
      user
    }

  override def resetUserSalt(): Future[UpdatedUserSalt] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.PUT,
          uri = s"$apiUrl/v1/users/self/salt"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      salt <- response.to[UpdatedUserSalt]
    } yield {
      salt
    }

  override def resetUserPassword(request: ResetUserPassword): Future[Done] =
    for {
      entity <- request.toEntity
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.PUT,
          uri = s"$apiUrl/v1/users/self/password",
          entity = entity
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      _ <- response.processed()
    } yield {
      Done
    }

  override def device(): Future[Device] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/devices/own/${self.toString}"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      device <- response.to[Device]
    } yield {
      device
    }

  override def pushDeviceKey(key: ByteString): Future[Done] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.PUT,
          uri = s"$apiUrl/v1/devices/own/${self.toString}/key"
        ).withEntity(key).addCredentials(credentials = credentials)
      ).transformClientFailures()
      _ <- response.processed()
    } yield {
      Done
    }

  override def pullDeviceKey(): Future[Option[ByteString]] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/devices/own/${self.toString}/key"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      key <- response match {
        case HttpResponse(status, _, _, _) if status.isSuccess() => response.toByteString.map(Some.apply)
        case HttpResponse(StatusCodes.NotFound, _, _, _)         => Future.successful(None)
        case _                                                   => response.asFailure
      }
    } yield {
      key
    }

  override def deviceKeyExists(): Future[Boolean] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.HEAD,
          uri = s"$apiUrl/v1/devices/own/${self.toString}/key"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      result <- response match {
        case HttpResponse(status, _, _, _) if status.isSuccess() => Future.successful(true)
        case HttpResponse(StatusCodes.NotFound, _, _, _)         => Future.successful(false)
        case _                                                   => response.asFailure
      }
    } yield {
      result
    }

  override def ping(): Future[Ping] =
    for {
      credentials <- credentials
      response <- offer(
        request = HttpRequest(
          method = HttpMethods.GET,
          uri = s"$apiUrl/v1/service/ping"
        ).addCredentials(credentials = credentials)
      ).transformClientFailures()
      ping <- response.to[Ping]
    } yield {
      ping
    }
}

object DefaultServerApiEndpointClient {
  import play.api.libs.json.Format

  def apply(
    apiUrl: String,
    credentials: => Future[HttpCredentials],
    decryption: DefaultServerApiEndpointClient.DecryptionContext,
    self: Device.Id,
    context: Option[EndpointContext],
    config: PoolClient.Config
  )(implicit system: ActorSystem[Nothing]): DefaultServerApiEndpointClient =
    new DefaultServerApiEndpointClient(
      apiUrl = apiUrl,
      credentials = credentials,
      decryption = decryption,
      self = self,
      context = context,
      config = config
    )

  sealed trait DecryptionContext

  object DecryptionContext {
    def apply(core: ServerCoreEndpointClient, deviceSecret: DeviceSecret, decoder: Decoder): DecryptionContext =
      Default(core = core, deviceSecret = deviceSecret, decoder = decoder)

    final case class Default(
      core: ServerCoreEndpointClient,
      deviceSecret: DeviceSecret,
      decoder: Decoder
    ) extends DecryptionContext

    case object Disabled extends DecryptionContext
  }

  implicit class ModelToRequestEntity[M](m: M) {
    import com.github.pjfanning.pekkohttpplayjson.PlayJsonSupport._

    def toEntity(implicit format: Format[M], ec: ExecutionContext): Future[RequestEntity] =
      Marshal(m).to[RequestEntity]
  }

  implicit class ExtendedResponseEntity(response: HttpResponse) {
    def processed[T](f: () => Future[T])(implicit system: ActorSystem[Nothing]): Future[T] = {
      import system.executionContext

      if (response.status.isSuccess()) {
        f()
          .recoverWith { case NonFatal(e) =>
            response.entity.dataBytes.cancelled().flatMap { _ =>
              Future.failed(
                new ServerApiFailure(
                  status = StatusCodes.InternalServerError,
                  message = s"Server API request unmarshalling failed with: [${e.getMessage}]"
                )
              )
            }
          }
      } else {
        response.asFailure
      }
    }

    def processed()(implicit system: ActorSystem[Nothing]): Future[Done] =
      processed[Done](f = () => Future.successful(Done))

    def to[M](implicit format: Format[M], ec: ExecutionContext, system: ActorSystem[Nothing]): Future[M] =
      processed[M] { () =>
        import com.github.pjfanning.pekkohttpplayjson.PlayJsonSupport._
        Unmarshal(response).to[M]
      }

    def toByteString(implicit ec: ExecutionContext, system: ActorSystem[Nothing]): Future[ByteString] =
      processed[ByteString] { () =>
        Unmarshal(response).to[ByteString]
      }

    def asFailure[T](implicit system: ActorSystem[Nothing]): Future[T] = {
      import system.executionContext

      Unmarshal(response)
        .to[String]
        .flatMap { responseContent =>
          Future.failed(
            new ServerApiFailure(
              status = response.status,
              message = s"Server API request failed with [${response.status.value}]: [$responseContent]"
            )
          )
        }
    }
  }

  implicit class HttpResponseWithTransformedFailures(op: => Future[HttpResponse]) {
    def transformClientFailures()(implicit ec: ExecutionContext): Future[HttpResponse] =
      op.recoverWith { case NonFatal(e: ClientFailure) =>
        Future.failed(
          new ServerApiFailure(
            status = StatusCodes.InternalServerError,
            message = e.getMessage
          )
        )
      }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/api/clients/DefaultServerApiEndpointClient.kt
package stasis.client_android.lib.api.clients

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.ByteString
import stasis.client_android.lib.api.clients.exceptions.ResourceMissingFailure
import stasis.client_android.lib.api.clients.internal.ClientExtensions
import stasis.client_android.lib.encryption.Decoder
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.server.api.requests.CreateDatasetDefinition
import stasis.client_android.lib.model.server.api.requests.CreateDatasetEntry
import stasis.client_android.lib.model.server.api.requests.ResetUserPassword
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetDefinition
import stasis.client_android.lib.model.server.api.responses.CreatedDatasetEntry
import stasis.client_android.lib.model.server.api.responses.Ping
import stasis.client_android.lib.model.server.api.responses.UpdatedUserSalt
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.model.server.devices.Device
import stasis.client_android.lib.model.server.devices.DeviceId
import stasis.client_android.lib.model.server.schedules.Schedule
import stasis.client_android.lib.model.server.schedules.ScheduleId
import stasis.client_android.lib.model.server.users.User
import stasis.client_android.lib.security.HttpCredentials
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.map
import stasis.client_android.lib.utils.Try.Companion.recoverWith
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success
import java.time.Instant

@Suppress("TooManyFunctions")
class DefaultServerApiEndpointClient(
    serverApiUrl: String,
    override val credentials: suspend () -> HttpCredentials,
    private val decryption: DecryptionContext,
    override val self: DeviceId
) : ServerApiEndpointClient, ClientExtensions() {
    override val server: String = serverApiUrl.trimEnd { it == '/' }

    override suspend fun datasetDefinitions(): Try<List<DatasetDefinition>> =
        jsonListRequest<DatasetDefinition> { builder ->
            builder
                .url("$server/v1/datasets/definitions/own")
                .get()
        }.map { definitions -> definitions.filter { it.device == self } }


    override suspend fun datasetDefinition(definition: DatasetDefinitionId): Try<DatasetDefinition> =
        jsonRequest<DatasetDefinition> { builder ->
            builder
                .url("$server/v1/datasets/definitions/own/${definition}")
                .get()
        }.map { actualDefinition ->
            if (actualDefinition.device == self) {
                actualDefinition
            } else {
                throw IllegalArgumentException("Cannot retrieve dataset definition for a different device")
            }
        }

    override suspend fun createDatasetDefinition(request: CreateDatasetDefinition): Try<CreatedDatasetDefinition> =
        if (request.device == self) {
            jsonRequest { builder ->
                builder
                    .url("$server/v1/datasets/definitions/own")
                    .post(request.toBody())
            }
        } else {
            Failure(
                IllegalArgumentException(
                    "Cannot create dataset definition for a different device: [${request.device}]"
                )
            )
        }

    override suspend fun datasetEntries(definition: DatasetDefinitionId): Try<List<DatasetEntry>> =
        jsonListRequest { builder ->
            builder
                .url("$server/v1/datasets/entries/own/for-definition/$definition")
                .get()
        }

    override suspend fun datasetEntry(entry: DatasetEntryId): Try<DatasetEntry> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/datasets/entries/own/$entry")
                .get()
        }

    override suspend fun latestEntry(
        definition: DatasetDefinitionId,
        until: Instant?
    ): Try<DatasetEntry?> {
        val baseUrl = "$server/v1/datasets/entries/own/for-definition/$definition/latest"

        return jsonRequest<DatasetEntry?> { builder ->
            builder
                .url(
                    when (until) {
                        null -> baseUrl
                        else -> "$baseUrl?until=$until"
                    }
                )
        }.recoverWith {
            when (it) {
                is ResourceMissingFailure -> Success(null)
                else -> Failure(it)
            }
        }
    }

    override suspend fun createDatasetEntry(request: CreateDatasetEntry): Try<CreatedDatasetEntry> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/datasets/entries/own/for-definition/${request.definition}")
                .post(request.toBody())
        }


    override suspend fun publicSchedules(): Try<List<Schedule>> =
        jsonListRequest { builder ->
            builder
                .url("$server/v1/schedules/public")
                .get()
        }


    override suspend fun publicSchedule(schedule: ScheduleId): Try<Schedule> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/schedules/public/$schedule")
                .get()
        }


    override suspend fun datasetMetadata(entry: DatasetEntryId): Try<DatasetMetadata> =
        when (val result = datasetEntry(entry)) {
            is Success -> datasetMetadata(entry = result.value)
            is Failure -> Failure(result.exception)
        }


    override suspend fun datasetMetadata(entry: DatasetEntry): Try<DatasetMetadata> =
        when (decryption) {
            is DecryptionContext.Default -> Try {

                val encryptedEntryMetadata = decryption.core.pull(crate = entry.metadata)

                DatasetMetadata.decrypt(
                    metadataCrate = entry.metadata,
                    metadataSecret = decryption.deviceSecret().toMetadataSecret(entry.metadata),
                    metadata = encryptedEntryMetadata,
                    decoder = decryption.decoder
                )
            }

            is DecryptionContext.Disabled -> {
                Failure(IllegalStateException("Cannot retrieve dataset metadata; decryption context is disabled"))
            }
        }

    override suspend fun user(): Try<User> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/users/self")
                .get()
        }

    override suspend fun resetUserSalt(): Try<UpdatedUserSalt> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/users/self/salt")
                .put("".toRequestBody())
        }

    override suspend fun resetUserPassword(request: ResetUserPassword): Try<Unit> =
        Try {
            request { builder ->
                builder
                    .url("$server/v1/users/self/password")
                    .put(request.toBody())
            }.successful()
        }

    override suspend fun device(): Try<Device> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/devices/own/$self")
                .get()
        }

    override suspend fun pushDeviceKey(key: ByteString): Try<Unit> =
        Try {
            request { builder ->
                builder
                    .url("$server/v1/devices/own/$self/key")
                    .put(key.toRequestBody("application/octet-stream".toMediaType()))
            }.successful()
        }

    override suspend fun pullDeviceKey(): Try<ByteString> =
        Try {
            val result = request { builder ->
                builder
                    .url("$server/v1/devices/own/$self/key")
                    .get()
            }.successful().body?.byteString()?.takeIf { it.size != 0 }

            when (result) {
                null -> throw ResourceMissingFailure()
                else -> result
            }
        }

    override suspend fun ping(): Try<Ping> =
        jsonRequest { builder ->
            builder
                .url("$server/v1/service/ping")
                .get()
        }

    sealed class DecryptionContext {
        data class Default(
            val core: ServerCoreEndpointClient,
            val deviceSecret: () -> DeviceSecret,
            val decoder: Decoder
        ) : DecryptionContext()

        data object Disabled : DecryptionContext()
    }
}
